File: 232.implement-queue-using-stacks.cpp
/*
 * @lc app=leetcode id=232 lang=cpp
 *
 * [232] Implement Queue using Stacks
 */

#include <bits/stdc++.h>

#include <cassert>

using namespace std;

// @lc code=start
class MyQueue {
 public:
  /** Initialize your data structure here. */
  MyQueue() {}

  /** Push element x to the back of queue. */
  void push(int x) {vec.push_back(x);}

  /** Removes the element from in front of queue and returns that element. */
  int pop() {
      int ret = peek();
      vec.erase(vec.begin());
      return ret;
  }

  /** Get the front element. */
  int peek() {return vec.front();}

  /** Returns whether the queue is empty. */
  bool empty() {return vec.empty();}

  private:
  vector<int> vec;
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */
// @lc code=end

int main() {
  MyQueue q;
  q.push(1);
  q.push(2);
  assert(!q.empty());
  assert(q.peek() == 1);
  assert(q.pop() == 1);
  assert(!q.empty());
  assert(q.peek() == 2);
  assert(q.pop() == 2);
  assert(q.empty());
  return 0;
}

File: 232.implement-queue-using-stacks.rs
/*
 * @lc app=leetcode id=232 lang=rust
 *
 * [232] Implement Queue using Stacks
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[macro_export]
macro_rules! list_node {
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let mut node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
struct MyQueue {
    vec: Vec<i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyQueue {
    fn new() -> Self {
        MyQueue { vec: vec![] }
    }

    fn push(&mut self, x: i32) {
        self.vec.push(x)
    }

    fn pop(&mut self) -> i32 {
        let ret = self.peek();
        self.vec.remove(0);
        ret
    }

    fn peek(&self) -> i32 {
        *self.vec.get(0).unwrap()
    }

    fn empty(&self) -> bool {
        self.vec.len() == 0
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * let obj = MyQueue::new();
 * obj.push(x);
 * let ret_2: i32 = obj.pop();
 * let ret_3: i32 = obj.peek();
 * let ret_4: bool = obj.empty();
 */
// @lc code=end

#[test]
fn test1() {
    let mut q = MyQueue::new();
    q.push(1);
    q.push(2);
    assert!(!q.empty());
    assert_eq!(q.peek(), 1);
    assert_eq!(q.pop(), 1);
    assert!(!q.empty());
    assert_eq!(q.peek(), 2);
    assert_eq!(q.pop(), 2);
    assert!(q.empty());
}
